web-app-pix2info-python/src/backend/docai.py

```diff
@@ -0,0 +1,278 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+import logging
+from io import BytesIO
+from mimetypes import guess_type
+from pathlib import Path
+from typing import Any, BinaryIO, Mapping, Sequence, TypeAlias, cast
+
+from google.cloud.documentai_v1 import (
+    Document,
+    DocumentProcessorServiceClient,
+    Processor,
+    ProcessRequest,
+    RawDocument,
+)
+from google.protobuf.json_format import MessageToJson  # type: ignore
+
+from .processors import (
+    DEMO_PROCESSING_LOCATIONS,
+    DEMO_PROCESSOR_TYPES,
+    SAMPLE_PROCESSING_LOCATION,
+    decode_processor_info,
+    encode_processor_info,
+)
+from .render import tiff_container_for_images, will_render_entity
+
+DocumentStream: TypeAlias = BytesIO
+MimeType: TypeAlias = str
+DocumentIO: TypeAlias = tuple[DocumentStream, MimeType]
+DocumentJson: TypeAlias = str
+DocumentData: TypeAlias = tuple[Document, DocumentJson]
+
+
+def get_client(location: str) -> DocumentProcessorServiceClient:
+    """Return a Document AI client."""
+    client_options = dict(api_endpoint=f"{location}-documentai.googleapis.com")
+    return DocumentProcessorServiceClient(client_options=client_options)
+
+
+def process_document(
+    file: BinaryIO,
+    mime_type: str,
+    project: str,
+    location: str,
+    processor_id: str,
+) -> Document:
+    """Analyze the input file with Document AI."""
+    client = get_client(location)
+    raw_document = RawDocument(content=file.read(), mime_type=mime_type)
+    name = client.processor_path(project, location, processor_id)
+
+    request = ProcessRequest(
+        raw_document=raw_document,
+        name=name,
+        skip_human_review=True,
+    )
+    response = client.process_document(request)
+
+    return response.document
+
+
+def process_documents(
+    documents: Sequence[DocumentIO],
+    project: str,
+    location: str,
+    processor_id: str,
+) -> DocumentData:
+    """Analyze the input file(s) with Document AI.
+
+    Return both the Document AI object and its json serialization (which can be viewed/cached).
+    """
+    if not documents:
+        raise ValueError("At least one document is expected")
+
+    if len(documents) == 1:
+        image_io, mime_type = documents[0]
+    else:
+        images = [file for (file, _) in documents]
+        image_io, mime_type = tiff_container_for_images(images)
+
+    document = process_document(image_io, mime_type, project, location, processor_id)
+    json = MessageToJson(
+        Document.pb(document),
+        preserving_proto_field_name=True,  # Same as Python (switch if camel case is preferred)
+        ensure_ascii=False,  # Don't escape characters (easier to read + smaller files)
+    )
```

Comment:
I don't think you need to use this extra library. You can use

```
Document.to_json(document)
```
to get a json string.

Reply:
Some additional explanations:
- It does not install an extra library: `google.protobuf` is a dependency of `google.cloud.documentai` (or any other client library).
- I went one level deeper with `google.protobuf.json_format.MessageToJson()` because `Document.to_json()` does not expose the `ensure_ascii` parameter (`to_json()` indirectly calls `MessageToJson()`, ... down to `json.dumps()`).
- I'm using the `ensure_ascii=False` parameter to make JSON serializations human-readable with non-ASCII text (and also files slightly smaller, though this is negligible). This is for educational reasons here (developers are able to properly read the json files with non-English text in their IDE).
- By default, non-ASCII characters are escaped. For example, this gives the following `receiver_name` for the French invoice:
  - `ensure_ascii=False`: `"mention_text": "MAËLLE ET JÉRÔME GUÉRIN"` (with a proper UTF-8 encoding)
  - `ensure_ascii=True`: `"mention_text": "MA\u00cbLLE ET J\u00c9R\u00d4ME GU\u00c9RIN"`

Reply:
OK, makes sense!